### PR TITLE
OpenVoice v2 voice-cloning pipeline + coached recording UI

### PR DESCRIPTION
**Pull request by HereLiesAz**
## Summary

Two-stage delivery of the OpenVoice v2 pipeline that replaces the old VITS speaker/acoustic/vocoder split. Targets users who **lost** their voice (ALS, laryngectomy, stroke, vocal cord injury) and want to keep sounding like themselves — for them, zero-shot cloning is the whole point, not a nice-to-have.

- **Tier 1** — `PocketTTSEngine.kt` rewritten end-to-end against the OpenVoice v2 contract (`pocket_tts_base.onnx` + `pocket_tts_se_extractor.onnx` + `pocket_tts_tone_converter.onnx`). 256-d speaker embeddings, 24 kHz output, multi-clip averaging.
- **Tier 2** — `VoiceCoachScreen` Compose UI: sequential clip recording with a live quality-tier indicator (●○○○ Initial → ●●●● Excellent) driven by an embedding-centroid stability argument.

### `PocketTTSEngine.kt`

Three ONNX sessions instead of four:
- `pocket_tts_base.onnx` — MeloTTS English base voice
- `pocket_tts_se_extractor.onnx` — 256-d speaker embedding extractor
- `pocket_tts_tone_converter.onnx` — timbre transfer (base → user)

Key behavior:
- `TTS_OUTPUT_SAMPLE_RATE_HZ` and `SE_EXTRACTOR_SAMPLE_RATE_HZ` both 24 000 (was 22 050 / 16 000).
- `EMBEDDING_DIM = 256` (was 192 for ECAPA-TDNN).
- `cloneVoice(name, files)` runs per-clip SE extraction and averages the 256-d embeddings. Empty list throws; dim mismatch across clips throws (so a stale 192-d profile never silently averages with a new 256-d one).
- `generateAudio(text, voice)`: tokenize → base TTS → lazy-cache base speaker SE → tone converter → user voice. Tokenizer returns null until a Kotlin `g2p_en` port lands, so `VoiceManager` falls back to system TTS; voice **cloning** still works end-to-end.
- WAV reader honors the embedded sample rate; 16/22.05/44.1/48 kHz reference clips (voicemails, video extracts, voice memos) are linearly resampled to 24 kHz before extraction. Multi-channel sources downmixed.

### Coached recording UI

- `VoiceQualityTier` — pure-Kotlin enum with `compute(clipCount, totalSeconds)`. Five bands (NONE → INITIAL → BASIC → GOOD → EXCELLENT) calibrated to where pure-averaging gains plateau against the embedding centroid.
- `CoachState` + five `VoiceViewModel` methods (`startCoachedSession`, `recordCoachedClip`, `discardLastCoachedClip`, `finishCoachedSession`, `cancelCoachedSession`). Clips live in `cacheDir`; engine resamples to 24 kHz at extraction.
- `VoiceCoachScreen` — script card cycling through Harvard Sentences (IEEE 269 phonetically balanced), big mic button (idle/recording/saving), live tier indicator with "next threshold" hint, captured-clips ribbon with discard-last, save-name dialog, discard-confirm dialog.
- `VoiceManagementScreen` gains a primary **Bank my voice (coached)** button above the existing import-files button.
- `LipertyApp` adds the `voice_coach` navigation destination.

### Supporting updates

- `setup_libs.sh` — `TTS_MODELS` list and FreeVC reference rewritten for the OpenVoice v2 file names.
- `ModelDownloadManager.kt` — three new `ModelSpec`s plus vocab JSON; the four legacy specs removed.
- `app/build.gradle.kts` — packaging excludes updated to the three new ONNX files so the AAB stays under the base-module size cap.
- `docs/VOICE_CLONER_STATE_2026-05-15.md` — full state writeup with what's blocked and the recovery path.

### Tests

- `PocketTTSEngineHelpersTest` — `linearResample` (identity / 2× / 0.5× / interp monotonicity), `averageEmbeddings` (single-clip / midpoint / 30-clip centroid convergence / dim-mismatch / empty-list), contract pins for 256-d and 24 kHz constants.
- `VoiceQualityTierTest` — zero clips, single-long-clip stays INITIAL (centroid stability requires multiple clips, not just more audio), threshold monotonicity, filled-dot monotonicity.
- `VoiceViewModelTest` extended with coach-state coverage (initial-empty, reset-after-cancel, save-with-zero-clips, discard-when-empty, defaults).

## What's blocked

1. **OpenVoice v2 ONNX exports** — `tools/export_tts_to_onnx.ipynb` (generator landed in commit `f4c17c1` on `main`) needs to be run on Kaggle to produce ~140 MB of artefacts on `HereLiesAz/liperty-pocket-tts`. Until then `setup_libs.sh` downloads 404 and the engine sessions are all null. The Compose UI gracefully falls back to system TTS in that state, so this PR is safe to merge before the notebook run completes.
2. **On-device tokenizer** — `g2p_en` Kotlin port pending. Until it lands `generateAudio()` returns null and `VoiceManager` uses system TTS. Voice **cloning** (record → embedding → save profile) works end-to-end without it.

## Test plan

- [ ] `./gradlew testDebugUnitTest` — verifies `PocketTTSEngineHelpersTest`, `VoiceQualityTierTest`, extended `VoiceViewModelTest` pass.
- [ ] Visual smoke test of `VoiceCoachScreen` on device: tier dots update after each clip, discard-last decrements the count, cancel dialog appears when there are captured clips, save dialog accepts a name and pops the screen.
- [ ] Visual smoke test of `VoiceManagementScreen`: new "Bank my voice (coached)" button navigates correctly; existing import wizard path still works.
- [ ] Once the Kaggle notebook lands the ONNX files, re-run `./setup_libs.sh` and verify clones produce non-null embeddings (256 floats per profile).

## Reviewer notes

- The Kotlin g2p_en port is a separate workstream (~500 LOC of CMU dict + rule fallbacks). Voice synthesis is intentionally gated off until that lands; this PR only delivers the cloning side.
- One uncommitted modification on `main` (`AvHubertEncoderSession.kt` — asset materialization fallback) was deliberately left out of this branch since it's pre-existing work unrelated to TTS.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Migrate the on-device TTS engine to the OpenVoice v2 zero-shot voice-cloning pipeline and introduce a coached multi-clip voice-banking flow with quality tier feedback, alongside supporting model download, packaging, and test updates.

New Features:
- Add OpenVoice v2-based PocketTTS engine with 24 kHz output and 256-d speaker embeddings, including multi-clip embedding averaging and robust WAV handling.
- Introduce a coached multi-clip voice-banking UI and state management, with live VoiceQualityTier progression and navigation from the voice management screen.

Enhancements:
- Update model download scripts, setup tooling, and Gradle packaging excludes to use the new OpenVoice v2 ONNX assets and vocab JSON.
- Document the current voice cloner state and migration path to the OpenVoice v2 pipeline.
- Add helper utilities and configuration for the new TTS pipeline, including readiness flags and synthesis gating on tokenizer availability.

Documentation:
- Add a detailed status document describing the OpenVoice v2 voice cloner pipeline, current limitations, and follow-up work.

Tests:
- Add unit tests for PocketTTSEngine helper functions (resampling, embedding averaging, and contract constants).
- Extend VoiceViewModel tests to validate coached recording state transitions and edge cases.
- Add tests for VoiceQualityTier thresholds, monotonicity, and filled-dot mapping.

### CODE REVIEW

File: app/src/main/java/com/hereliesaz/liperty/voicebox/VoiceViewModel.kt [hunk 1]
```diff
@@ -330,6 +371,174 @@ class VoiceViewModel(application: Application) : AndroidViewModel(application) {
...
+        )
+    }
+
+    /**
+     * Reads a 16-bit PCM WAV's data-chunk length without decoding
+     * the audio. Falls back to length/2/16000 if the header isn't
+     * a standard 44-byte RIFF.
+     */
+    private fun wavDurationMs(file: File): Long {
+        val bytes = try { file.readBytes() } catch (_: Exception) { return 0L }
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/voicebox/VoiceViewModel.kt:519]: VoiceViewModel.wavDurationMs now reads only the WAV header via the new shared WavHeader.read() streaming util (a few hundred bytes), with a length-based fallback. Fixed in #99.

File: app/src/main/java/com/hereliesaz/liperty/voicebox/cloning/VoiceQualityTier.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+         * Computes the highest tier a profile with [clipCount] clips
+         * and [totalSeconds] cumulative speech qualifies for.
+         *
+         * Both thresholds must be met: a single 60-second clip is
+         * still [INITIAL] because the centroid is dominated by one
+         * estimate (the "more clips → more centroid stability"
+         * argument requires multiple clips, not just more audio).
+         */
+        fun compute(clipCount: Int, totalSeconds: Int): VoiceQualityTier {
+            val candidates = values().filter { tier ->
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/voicebox/cloning/VoiceQualityTier.kt]: Done — switched values() to entries (and added a rationale comment). The maintainer applied this directly via 5e1ea00; updated the test file to match in #99.

File: app/src/main/java/com/hereliesaz/liperty/ui/VoiceCoachScreen.kt [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+
+    if (showCancelConfirm) {
+        AlertDialog(
+            onDismissRequest = { showCancelConfirm = false },
+            title = { Text("Discard recordings?") },
+            text = {
+                Text(
+                    "${state.clipCount} clip(s) will be deleted. " +
+                        "Your voice profile won't be saved."
+                )
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/ui/VoiceCoachScreen.kt:112]: Added voice_coach_clips_captured and voice_coach_clip_count_inline plurals to strings.xml; VoiceCoachScreen routes the captured-clip banner + inline tier counter through pluralStringResource. Fixed in #99.

File: app/src/main/java/com/hereliesaz/liperty/voicebox/PocketTTSEngine.kt [hunk 4]
```diff
@@ -7,322 +7,605 @@ import android.content.Context
...
+     * Minimal WAV reader (16-bit PCM mono). Honors the sample rate
+     * declared in the header so a user-supplied 16/22.05/44.1/48 kHz
+     * clip (voicemail, video extract, voice memo) can be resampled
+     * to the 24 kHz the extractor expects.
+     *
+     * Falls back to assuming 24 kHz mono PCM if the header is
+     * unparseable, matching the previous behavior.
+     */
+    private fun readWavFile(file: File): Pair<FloatArray, Int> {
+        val bytes = file.readBytes()
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/voicebox/PocketTTSEngine.kt:329]: PocketTTSEngine.readWavFile now reads the header via WavHeader.read() then streams the data chunk via a 32 KB buffer, downmixing to mono on the fly — no full-file ByteArray. Hard-capped at 40 MB per clip. Fixed in #99.

File: app/src/main/java/com/hereliesaz/liperty/ui/VoiceCoachScreen.kt [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+        // ── Quality tier card ────────────────────────────────────
+        QualityTierCard(
+            tier = state.qualityTier,
+            clipCount = state.clipCount,
+            totalSeconds = state.totalSpeechSeconds,
+        )
+
+        // ── Script card ──────────────────────────────────────────
+        val script = remember(state.clipCount) {
+            HARVARD_PROMPTS[state.clipCount % HARVARD_PROMPTS.size]
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/ui/VoiceCoachScreen.kt:161]: Added promptIndex to CoachState; recomputeCoachQualityState advances it only on successful capture (advancePrompt = true). Discards leave promptIndex alone, so retrying lands on the same sentence. Fixed in #99.

File: app/src/main/java/com/hereliesaz/liperty/voicebox/VoiceViewModel.kt [hunk 6]
```diff
@@ -330,6 +371,174 @@ class VoiceViewModel(application: Application) : AndroidViewModel(application) {
...
+        _coachState.value = _coachState.value.copy(
+            isRecording = true,
+            statusMessage = "Recording…"
+        )
+        voiceRecorder.startRecording(
+            durationMs = durationMs,
+            onComplete = { file, _ ->
+                // VoiceRecorder writes 16 kHz mono. The WAV header
+                // carries the rate, so PocketTTSEngine.readWavFile()
+                // resamples to 24 kHz at extraction time and
```
[the remark this answers is not in the document]
Comment (HereLiesAz) [reply app/src/main/java/com/hereliesaz/liperty/voicebox/VoiceViewModel.kt:409]: recomputeCoachQualityState now explicitly sets error = null on every success path. Fixed in #99.
